pkg/globalconfig/global_config.go

```diff
@@ -155,3 +166,74 @@ func GetValidOmitContainers() []string {
 
 	return s
 }
+
+// HostPortIsAllocated returns the project name that has allocated
+// the port, or empty string.
+func HostPostIsAllocated(port string) string {
+	for project, item := range DdevGlobalConfig.ProjectList {
+		if nodeps.ArrayContainsString(item.UsedHostPorts, port) {
+			return project
+		}
+	}
+	return ""
+}
+
+// Check GlobalDdev UsedHostPorts to see if requested ports are available.
+func CheckHostPortsAvailable(projectName string, ports []string) error {
+	for _, port := range ports {
+		allocatedProject := HostPostIsAllocated(port)
+		if allocatedProject != projectName && allocatedProject != "" {
+			return fmt.Errorf("host port %s has already been allocated to project %s", port, allocatedProject)
+		}
+	}
+	return nil
+}
+
+// GetFreePort gets an ephemeral port currently available, but also not
+// listed in DdevGlobalConfig.UsedHostPorts
+func GetFreePort(localIPAddr string) (string, error) {
+	// Limit tries arbitrarily. It will normally succeed on first try.
+	for i := 1; i < 1000; i++ {
+		// From https://github.com/phayes/freeport/blob/master/freeport.go#L8
+		// Ignores that the actual listener may be on a docker toolbox interface,
+		// so this is just a heuristic.
+		addr, err := net.ResolveTCPAddr("tcp", "127.0.0.1:0")
+		if err != nil {
+			return "", err
+		}
+
+		l, err := net.ListenTCP("tcp", addr)
+		if err != nil {
+			return "", err
+		}
+		port := strconv.Itoa(l.Addr().(*net.TCPAddr).Port)
+		// nolint: errcheck
+		l.Close()
+
+		// In the case of Docker Toolbox, the actual listening IP may be something else
+		// like 192.168.99.100, so check that to make sure it's not currently occupied.
+		conn, _ := net.Dial("tcp", localIPAddr+":"+port)
+		if conn != nil {
+			continue
+		}
+
+		if HostPostIsAllocated(port) != "" {
+			continue
+		}
+		return port, nil
+	}
+	return "-1", fmt.Errorf("GetFreePort() failed to find a free port")
+
+}
+
+// ReservePorts() adds the ProjectInfo if necessary and assigns the reserved ports
+func ReservePorts(projectName string, ports []string) error {
+	// If the project doesn't exist, add it.
+	_, ok := DdevGlobalConfig.ProjectList[projectName]
```

Comment:
After playing around with this post-update, I think there should be a check either here or perhaps in the caller that returns early if `len(ports) == 0`. As written, an empty `ProjectInfo` struct and `usedhostports` list are written to global_config.yaml, even if no ports are reserved -- not the end of the world, but adds some unneeded clutter to that config file.

Reply:
This sounded like a great idea, and I started working on it... but since I was a good boy and used a struct instead of a map, I can't do it without removing the entire ProjectInfo, which would mess up future uses of the list.